llama-index-core/llama_index/core/memory/chat_summary_memory_buffer.py

```diff
@@ -0,0 +1,245 @@
+from typing import Any, Callable, Dict, List, Optional
+
+from pydantic.fields import PrivateAttr
+
+from llama_index.core.base.llms.types import ChatMessage, MessageRole
+from llama_index.core.bridge.pydantic import Field, root_validator
+from llama_index.core.llms.llm import LLM
+from llama_index.core.memory.types import DEFAULT_CHAT_STORE_KEY, BaseMemory
+from llama_index.core.storage.chat_store import BaseChatStore, SimpleChatStore
+from llama_index.core.utils import get_tokenizer
+
+DEFAULT_TOKEN_LIMIT_FULL_TEXT_RATIO = 0.75
+DEFAULT_TOKEN_LIMIT_FULL_TEXT = 2000
+SUMMARIZE_PROMPT = "The following is a conversation between the user and assistant. Write a concise summary about the contents of this conversation."
+
+
+# TODO: Add option for last N user/assistant history interactions instead of token limit

```

Comment:
side note, slightly off topic but: does N messages make sense? What if a single user or assistant message is huge? Also another consideration, for function calls, openai (and others) require pairs of an assistant and tool message to be in memory

Reply:
If a single user or assistant message is huge, then indeed, this is not such a great idea. On the other hand, the user has knowledge about the prompts and the expected length of the responses (perhaps something along the lines of "Answer in a single sentence: what is LlamaIndex?") and it could be a use case to consider only the last 3 or 5 messages, say.

The reasons I considered it is to ensure that USER/ASSISTANT messages are always considered in pairs, and to give some more flexibility to the end user. But it's more of a nice to have for the future if there is appetite for that flexibility.